### PR TITLE
Normalize table

### PR DESCRIPTION
**Pull request by sowe9385**
fix comments of @gregcaporaso and @jairideout

### CODE REVIEW

**Comment by qiime-r2d2** [comment]
Build results will soon be (or already are) available at: http://ci.qiime.org/job/qiime-github-pr/1487/

**Comment by jairideout** [comment]
retest this please

**Comment by qiime-r2d2** [comment]
Build results will soon be (or already are) available at: http://ci.qiime.org/job/qiime-github-pr/1488/

**Comment by qiime-r2d2** [comment]
Build results will soon be (or already are) available at: http://ci.qiime.org/job/qiime-github-pr/1490/

**Comment by qiime-r2d2** [comment]
Build results will soon be (or already are) available at: http://ci.qiime.org/job/qiime-github-pr/1492/

**Comment by jairideout** [comment]
Thanks @sowe9385! Most of my comments in #1820 are relevant to this pull request. In addition, can you please add a new unit test to catch the bug we found (this is the third item in #1810)?

**Comment by qiime-r2d2** [comment]
Build results will soon be (or already are) available at: http://ci.qiime.org/job/qiime-github-pr/1498/

**Comment by antgonza** [comment]
retest this please

**Comment by qiime-r2d2** [comment]
Build results will soon be (or already are) available at: http://ci.qiime.org/job/qiime-github-pr/1499/

File: qiime/normalize_table.py [hunk 1]
```diff
@@ -20,16 +24,18 @@
...
 
     if output_CSS_statistics:
         base_fname, ext = splitext(out_path)
         output_CSS_statistics = base_fname+'_CSS_statistics.txt'
-    
-    run_CSS(json_infile, out_path, output_CSS_statistics=output_CSS_statistics)
-    remove(json_infile)
+
+    with tempfile.NamedTemporaryFile(dir=get_qiime_temp_dir(),
+                                 prefix='QIIME-differential-abundance-temp-table-',
```
Comment (jairideout) [inline qiime/normalize_table.py]: Please update the prefix here and in the rest of the code to be `QIIME-normalize-table-temp-table-`.

File: qiime/normalize_table.py [hunk 2]
```diff
@@ -42,20 +48,22 @@ def multiple_file_normalize_CSS(input_dir, output_dir, output_CSS_statistics):
...
-        remove(json_infile)
+
+        with tempfile.NamedTemporaryFile(dir=get_qiime_temp_dir(),
+                                     prefix='QIIME-differential-abundance-temp-table-',
+                                     suffix='.biom') as temp_fh:
+            temp_fh.write(tmp_bt.to_json('forR'))
+            temp_fh.flush()
+            run_CSS(temp_fh.name, outfile, output_CSS_statistics=output_CSS_statistics)
 
 def run_CSS(input_path, out_path, output_CSS_statistics, HALT_EXEC=False):
```
Comment (jairideout) [inline qiime/normalize_table.py]: `HALT_EXEC` isn't used, can you please remove?

Comment (jairideout) [inline qiime/normalize_table.py]: Also remove in the other functions in this file

File: qiime/normalize_table.py [hunk 3]
```diff
@@ -42,20 +48,22 @@ def multiple_file_normalize_CSS(input_dir, output_dir, output_CSS_statistics):
...
-        open(str(json_infile),'w').write(load_table(hdf5_infile).to_json('forR'))
+        tmp_bt = load_table(hdf5_infile) 
         outfile = join(output_dir, 'CSS_'+base_fname+'.biom')
         if output_CSS_statistics:
             output_CSS_statistics = join(output_dir, 'CSS_statistics_'+base_fname+'.txt')
-        run_CSS(json_infile, outfile, output_CSS_statistics=output_CSS_statistics)
-        remove(json_infile)
+
+        with tempfile.NamedTemporaryFile(dir=get_qiime_temp_dir(),
+                                     prefix='QIIME-differential-abundance-temp-table-',
```
Comment (jairideout) [inline qiime/normalize_table.py]: fix indentation on these two lines (and in the other functions where this occurs)

**Comment by jairideout** [comment]
@sowe9385 I had a few more inline comments (all minor). In addition, can you please take care of the following?

- [ ] use `option_parser.error` to fail with an appropriate error message when a "required" option isn't present. See my comment [here](https://github.com/biocore/qiime/pull/1820#issuecomment-68961432) for more details.
- [ ] add missing unit test (see my comment [here](https://github.com/biocore/qiime/pull/1821#issuecomment-68792625) for more details)

**Comment by qiime-r2d2** [comment]
Build results will soon be (or already are) available at: http://ci.qiime.org/job/qiime-github-pr/1501/

File: scripts/normalize_table.py [hunk 4]
```diff
@@ -99,6 +109,11 @@ def main():
     algorithm = opts.algorithm
     list_algorithms = opts.list_algorithms
 
+    almost_required_options = ['input_path', 'out_path']
```
Comment (jairideout) [inline scripts/normalize_table.py]: Same comment here as with differential_abundance.py: running `normalize_table.py -l` raises an error.

File: tests/test_normalize_table.py [hunk 5]
```diff
@@ -61,9 +61,9 @@ def test_normalize_table_CSS(self):
             load_table(self.tmp_otu_fp_out_CSS).ids())
 
         #test taxonomy added to CSS; DESeq gives negatives so no taxonomy added
-        # self.assertItemsEqual(
-        #     q._observation_metadata,
-        #     load_table(self.tmp_otu_fp_out_CSS)._observation_metadata)
+        self.assertItemsEqual(
+            q._observation_metadata,
```
Comment (jairideout) [inline tests/test_normalize_table.py]: use `q.metadata(axis='observation')` here and on the next line

**Comment by jairideout** [comment]
retest this please

**Comment by qiime-r2d2** [comment]
Build results will soon be (or already are) available at: http://ci.qiime.org/job/qiime-github-pr/1503/

**Comment by qiime-r2d2** [comment]
Build results will soon be (or already are) available at: http://ci.qiime.org/job/qiime-github-pr/1509/

**Comment by qiime-r2d2** [comment]
Build results will soon be (or already are) available at: http://ci.qiime.org/job/qiime-github-pr/1511/

**Comment by jairideout** [comment]
Thanks @sowe9385!

**Pull request merged**